server/lib/transferwise.ts

```diff
@@ -90,55 +90,104 @@ const parseError = (
+  const tokenCreation = moment.utc(connectedAccount.data.created_at);
+  const diff = moment.duration(moment.utc().diff(tokenCreation)).asSeconds();
+  const isOutdated = diff > <number>connectedAccount.data.expires_in - 60;
+  if (refresh || isOutdated) {
+    const newToken = await getOrRefreshToken({ refreshToken: connectedAccount.refreshToken });
+    if (!newToken) {
+      throw new Error('There was an error refreshing the Transferwise token');
+    }
+    const { access_token: token, refresh_token: refreshToken, ...data } = newToken;
+    await connectedAccount.update({ token, refreshToken, data: { ...connectedAccount.data, ...data } });
+    return token;
+  } else {
+    return connectedAccount.token;
+  }
+}
+
+export const requestDataAndThrowParsedError = async (
+  fn: (url, data?, options?) => Promise<AxiosResponse>,
   url: string,
   {
     data,
     requestPath,
+    connectedAccount,
+    retries = 0,
     ...options
   }: {
     data?: Record<string, unknown>;
     headers?: Record<string, unknown>;
     params?: Record<string, unknown>;
     auth?: Record<string, unknown>;
     requestPath?: string;
+    connectedAccount?: ConnectedAccount;
+    retries?: number;
   },
   defaultErrorMessage?: string,
 ): Promise<any> => {
   const start = process.hrtime.bigint();
-  debug(`calling ${config.transferwise.apiUrl}${url}: ${JSON.stringify({ data, params: options.params }, null, 2)}`);
-  const pRequest = data ? fn(url, data, options) : fn(url, options);
-  return pRequest
-    .then(getData)
-    .then(tap(data => debug(JSON.stringify(data, null, 2))))
-    .catch(e => {
-      // Implements Strong Customer Authentication
-      // https://api-docs.transferwise.com/#payouts-guide-strong-customer-authentication
-      const signatureFailed = e?.response?.headers['x-2fa-approval-result'] === 'REJECTED';
-      const hadSignature = e?.response?.headers['X-Signature'];
-      if (signatureFailed && !hadSignature) {
-        const ott = e.response.headers['x-2fa-approval'];
-        const signature = signString(ott);
-        options.headers = { ...options.headers, 'X-Signature': signature, 'x-2fa-approval': ott };
-        const request = data ? fn(url, data, options) : fn(url, options);
-        return request.then(getData);
-      } else {
-        throw e;
-      }
-    })
-    .catch(e => {
+
+  if (connectedAccount) {
+    const token = await getToken(connectedAccount);
+    set(options, 'headers.Authorization', `Bearer ${token}`);
+  }
+
+  debug(
+    `calling ${config.transferwise.apiUrl}${url}: ${JSON.stringify(
+      { data, params: options.params, retries },
+      null,
+      2,
+    )}`,
+  );
+
+  try {
+    const pRequest = data ? fn(url, data, options) : fn(url, options);
+    const response = await pRequest;
+    return getData(response);
+  } catch (e: any) {
+    const signatureFailed = e?.response?.headers?.['x-2fa-approval-result'] === 'REJECTED';
+    const hadSignature = e?.response?.headers?.['X-Signature'];
```

Comment:
```suggestion
    const hadSignature = options?.headers?.['X-Signature'];
```

looking at what we sent in the request to decide to recurse or not.